libs/vkd3d/device.c

```diff
@@ -5106,17 +5121,109 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_OpenSharedHandleByName(d3d12_devic
 static HRESULT STDMETHODCALLTYPE d3d12_device_MakeResident(d3d12_device_iface *iface,
         UINT object_count, ID3D12Pageable * const *objects)
 {
-    FIXME_ONCE("iface %p, object_count %u, objects %p stub!\n",
-            iface, object_count, objects);
+    struct d3d12_device *device = impl_from_ID3D12Device(iface);
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+
+    if (device->vk_info.EXT_pageable_device_local_memory)
+    {
+        uint32_t i;
+        
+        for (i = 0; i < object_count; i++)
+        {
+            ID3D12Heap *heap_iface;
+            ID3D12Resource *resource_iface;
+            ID3D12DeviceChild *object = (ID3D12DeviceChild*)objects[i];
+            VkDeviceMemory memory = VK_NULL_HANDLE;
+            D3D12_RESIDENCY_PRIORITY priority;
+
+            if (SUCCEEDED(ID3D12DeviceChild_QueryInterface(object, &IID_ID3D12Heap, (void**)&heap_iface)))
+            {
+                struct d3d12_heap *heap_object = impl_from_ID3D12Heap(heap_iface);
+
+                memory = heap_object->allocation.device_allocation.vk_memory;
+                priority = heap_object->priority;
+                heap_object->evicted = false;
+
+                ID3D12Heap_Release(heap_iface);
+            }
+            else if (SUCCEEDED(ID3D12DeviceChild_QueryInterface(object, &IID_ID3D12Resource, (void**)&resource_iface)))
+            {
+                struct d3d12_resource *resource_object = impl_from_ID3D12Resource(resource_iface);
+
+                memory = (resource_object->flags & VKD3D_RESOURCE_COMMITTED) ?
+                    resource_object->mem.device_allocation.vk_memory : VK_NULL_HANDLE;
+                priority = resource_object->priority;
+                resource_object->evicted = false;
```

Comment:
Is this a potential race? I'm not sure if it's valid to call MakeResident concurrently on the same resource.

Reply:
Also worth noting that docs say:

```
MakeResident is ref-counted, such that [Evict](https://learn.microsoft.com/en-us/windows/desktop/api/d3d12/nf-d3d12-id3d12device-evict) must be called the same amount of times as MakeResident before Evict takes effect. Objects that support residency are made resident during creation, so a single Evict call will actually evict the object.
```

So the appropriate implementation should probably use atomic ref-count and priority is atomic relaxed exchange, where initial ref-count is 1 or 0 depending on the heap flag.

Reply:
Alternatively, to stay sane, a simple lock per-object around the residency processing is fine too.

Reply:
I've now added a spinlock around the residency fields.  Although the actual vkSetDeviceMemoryPriorityEXT() isn't covered by the lock, I think I've convinced myself that it should be (in which case that spinlock should really be a mutex). :(

Reply:
Residency-counting is in, now.